### ISSUE TITLE
No testing of database migrations from previous releases

### ISSUE THREAD

**Issue by fxthomas** [airsonic/airsonic#1363]
This is a sibling issue of #1351, but probably easier to do.

We should add an empty (but fully migrated) database for each of the previous versions in our tests, including Subsonic and Libresonic, so that we have minimal coverage for Airsonic upgrades.

**Comment by muff1nman** [comment]
@randomnicode reminded me that we have one existing example which is the `org.airsonic.player.service.LegacyDatabaseStartupTestCase`. We probably just need to put some work into generalizing it and making sure we add migrated databases for each release. We could also get fancy and move it to the docker test harness and test with the different databases we support.

**Comment by randomnicode** [comment]
Yeah, I think that test is buggy and isn't testing the right db, though it is meant for it.

I don't know if we publish our artifacts (on merge) somewhere, but two ideas are:
- At git level
   - Get travis to run a migration of master prior to PR merge (either fetch master, or perhaps travis just gives you the master branch), this establishes baseline
   - Run migrations/start app with the PR merged into master, pointed at the same DB. This should see if the new migrations in the PR are working with the old ones
   - Pros: no external artifacts required
   - Cons: need to finagle with travis and git branches

- At DB artifact level (more common)
    - Persist DB artifact produced at each merge. Tag it (as 'latest' or whatever)
    - Include a test that fetches the 'latest' DB and runs the PR migrations on it and makes sure everything goes smoothly
    - Pros: Pretty standard pattern
    - Cons: Needs an artifactory/hub/repo, some place to put the db artifacts. Relies on a previous migration (so someone will have to upload a starting db to the right location the first time around).

I would actually do it at each merge to master instead of release.

**Comment by fxthomas** [comment]
I was thinking of the second solution.

Also, with HSQLDB the empty databases are actually text files _and_ fairly small (since they are just an SQL script and a `.properties` file). Why not bundle them in our Git repository?

**Comment by muff1nman** [comment]
Yeah we should probably just bundle them in with the repo. We already have git lfs enabled and can track them using lfs if they're larger than a MB.

**Comment by randomnicode** [comment]
Hmm, I'm not sure what "bundling with the repo" means here.

Are we saying that every time we check in a migration, we should run it on a local system and add the resulting script file in _addition_ to the migration? That sounds a little kludgy to me. We kind of do it with the pre-liquibase files, and I guess I understand for compatibility, but...

For one, it doesn't help test dbs other than HSQLDB. And by going down that line of reasoning, we can say, why not check in Postgres and My/MariaDB files also? At the end of the day, _all_ dbs are just files.
For two, anyone can modify the script to "massage" the migration and check it in. No one would know how it was generated.

Just my two cents.

**Comment by fxthomas** [comment]
> Hmm, I'm not sure what "bundling with the repo" means here.

Committing them in the repository as is.

> every time we check in a migration

I was thinking more along the lines of "for each minor release". It avoids _accidentally_ breaking a release while limiting the overhead of having to commit each time we do a migration, and as a bonus it lets us test database version upgrades (like going from HSQLDB 1.8 to 2.5).

Also, I am wary of fiddling with Travis to make it correctly run migrations on `master` and _then_ on the PR build each time we do a PR. I could be wrong, but for me it would take much more time to implement correctly than the solution I am describing.

> For two, anyone can modify the script to "massage" the migration and check it in.

This of course trusts that nobody will _non-accidentally_ edit the database scripts by hand to force tests to pass. But all test cases suffer from this issue.

> For one, it doesn't help test dbs other than HSQLDB.

For other DBs, unfortunately, you're right. I don't see an easy way of testing intermediate migrations, which is why I opened a more limited version of this in #1351.
